daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala

```diff
@@ -1806,68 +1881,101 @@ object VerifyTestCase {
+      actualDiags.filter(diag => diag.isValidation)
+    } else if (expectedDiagnosticType == "warning") {
+      actualDiags.filterNot(d => d.isValidation || d.isError)
+    } else if (expectedDiagnosticType == "error") {
+      actualDiags.filter(_.isError)
+    } else {
+      Seq.empty
+    }
+
+    val actualDiagsFilteredMessages = actualDiagsFiltered.map(_.toString())
+
+    // throw exception if we have no actualDiags, but have expected diagnostics
+    if (
+      expectedDiags.isDefined
+      && actualDiagsFiltered.isEmpty
+    ) {
       throw TDMLException(
         """"Diagnostic message(s) were expected but not found."""" +
           "\n" +
           """Expected: """ + expectedDiagMsgs.mkString("\n") +
-          (if (actualDiagMsgs.isEmpty)
-             "\n No diagnostic messages were issued."
+          (if (actualDiagsFiltered.isEmpty)
+             s"\n No ${expectedDiagnosticType} diagnostic messages were issued."
            else
-             "\n The actual diagnostics messages were: " + actualDiagMsgs.mkString("\n")),
+             "\n The actual diagnostics messages were: " + actualDiagsFilteredMessages.mkString(
+               "\n"
+             )),
         implString
       )
     }
 
-    // must find each expected warning message within some actual warning message.
-    expectedDiags.foreach { expectedDiag =>
-      {
-        expectedDiag.messages.foreach { expectedMsg =>
-          {
-            val wasFound = actualDiags.exists { actualDiag =>
-              val actualDiagMsg = actualDiag.getMessage()
-              actualDiag.isError == expectedDiag.isError &&
-              actualDiagMsg.toLowerCase.contains(expectedMsg.toLowerCase)
-            }
-            if (!wasFound) {
-              throw TDMLException(
-                s"""Did not find diagnostic ${expectedDiag.diagnosticType} message """" +
-                  expectedMsg +
-                  """" in any of the actual diagnostic messages: """ + "\n" +
-                  actualDiagMsgs.mkString("\n"),
-                implString
-              )
-            }
+    if (
+      !actualDiags.forall(d => d.isValidation || d.isError)
+      && expectedDiags.isEmpty
+      && !ignoreUnexpectedWarnings
+    ) {
+      throw TDMLException(
+        s"ignoreUnexpectedWarnings = false and test does not expect warnings, but created the following: " +
+          s"${actualDiags.filterNot(d => d.isValidation || d.isError).mkString("\n")}",
+        implString
+      )
+    }
+
+    expectedDiags.map { diags =>
+      diags.map { diag =>
+        val matchAttr = diag.matchAttrib
+        val diagsFound = diag.messages.map { d =>
+          // make case insensitive
+          val wasFound =
+            actualDiagsFilteredMessages.exists(ad => ad.toLowerCase.contains(d.toLowerCase))
+          (d, wasFound)
+        }
+        matchAttr match {
+          case "all" if (!diagsFound.forall(_._2)) => {
```

Comment:
Technically, you probably want the condition inside the case block. For example, currently if matchAttr is "all" but the condition fails then it will do the "none" and efault cases, which shouldn't happen. And that way, the default case implies the user had something like `match="foo"` and it should be an error.